### PR TITLE
Create DefaultsGuidance.md

### PR DESCRIPTION
**Pull request by markwilkie**
Initial DRAFT for guidance on how to think about 'defaults' and Arcade

### CODE REVIEW

**Comment by natemcmaster** [comment]
Thanks for writing this up. I would add concrete examples of each category. Otherwise, LGTM!

**Comment by markwilkie** [comment]
Also added some contributor guidance as it's often relevant....

File: Documentation/ArcadeContributorGuidance.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+For the most part, contributions to Arcade are straight forward and relatively 'smooth'.  However, from time to time, getting changes we want in can definitely be more challenging, and even frustrating.  This document attempts to clarify some of the expectations as well as provide some 'advice' for success when contributing to Arcade.
+
+## Principles
+* Should benefit multiple repos across .NET Core
+* Keep it simple (don't over engineer or be "clever")
+* Focus on value over conviction.  (of course conviction should help drive the value discussion)
+* Opinions and beliefs must be substantiated with data and a strong business rationale
+* All discussions (arguments) must be civil, respectful, productive, and above all - kind. The focus must always be on the topic, not individuals.
+* Work in 'partnership' (shoulder to shoulder) as apposed to 'us' vs. 'them'.
+
```
Comment (jcagme) [inline Documentation/ArcadeContributorGuidance.md]: Nit: Some lines end with '.' and some others don't. Make all end up in the same way

File: Documentation/ArcadeContributorGuidance.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+## Conflict Resolution / Arbitration
+What happens when there's just no discernable path forward?  Who is the final arbitrator?  What to do in those rare occasions where everyone is stuck....
```
Comment (jcagme) [inline Documentation/ArcadeContributorGuidance.md]: >.... [](start = 149, length = 4)

.... -> ....

File: Documentation/ArcadeContributorGuidance.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+* Keep it simple (don't over engineer or be "clever")
+* Focus on value over conviction.  (of course conviction should help drive the value discussion)
+* Opinions and beliefs must be substantiated with data and a strong business rationale
+* All discussions (arguments) must be civil, respectful, productive, and above all - kind. The focus must always be on the topic, not individuals.
+* Work in 'partnership' (shoulder to shoulder) as apposed to 'us' vs. 'them'.
+
+## Conflict Resolution / Arbitration
+What happens when there's just no discernable path forward?  Who is the final arbitrator?  What to do in those rare occasions where everyone is stuck....  
+
+The short answer is that the final arbitrator is the Arcade owner, which currently is @markwilkie.  However....before jumping in and asking for a 'judgement', there are a few things that need to happen first.
```
Comment (jcagme) [inline Documentation/ArcadeContributorGuidance.md]: >.... [](start = 107, length = 4)

.... -> ....

Comment (jcagme) [inline Documentation/ArcadeContributorGuidance.md]: >. [](start = 207, length = 1)

. -> :

File: Documentation/ArcadeContributorGuidance.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+* Let **time** go by, while continuing to poke, discuss, re-think and so on.  Most things don't need resolution urgently (even though we often feel urgent about them) and usually the right path forward gets more clear over time.  In cases where it is urgent, finding a short term 'hack' is perfectly acceptable.
+* Keep channels of communication open, and stay in there with tenacity.  In disagreements which are tough, if the dev pushing for a change loses heart, then likely nothing will happen.  Stay at it and keep up the pressure.  This is often tough to do...
+* Have at least three (3) face-to-face conversations (or voice when remote) over a week or two at a minimum with the devs where there is the most disagreement.  It is amazing how much these higher bandwidth conversations help.
```
Comment (jcagme) [inline Documentation/ArcadeContributorGuidance.md]: >Have at least three (3) face-to-face conversations (or voice when remote) [](start = 2, length = 73)

As a general OS question, nor specific to Arcade, what's the recommended in-person channel of communication (other than GitHub) for contributors outside of Msft?

File: Documentation/ArcadeContributorGuidance.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+* Be able to articulate all other points of view.  Know the rationale behind the differences.
+
+Time and again differences get worked out when these items occur.  To be sure, it's not easy and we're often tempted to think "it shouldn't be this hard".  While perhaps that's true - the reality is that sometimes it IS this hard, and there's really no easy path through.  The good news is that in Arcade, this is rare.
+
+## Helpful Tactics/Advice when it's challenging
+* Where there are differences that are challenging, see the 'Defaults Guidance' doc for help on framing and definitions.
```
Comment (jcagme) [inline Documentation/ArcadeContributorGuidance.md]: >Defaults Guidance [](start = 61, length = 17)

Make this a link to DefaultGuidance.md

File: Documentation/DefaultsGuidance.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+# Defaults and Arcade
+
+It is often challenging to reconcile differences between a given repo build and Arcade - which represents .NET Core overall.  It's sometime tough to determine when to conform to how Arcade expects settings/properties/patterns/targets/conventions to be and when to 'override' (fork) for a given repo.  This document attempts to provide some guidance which hopefully gives at least a place to start - as well as a 'frame' for discussions.
+
+## Reminders/Principles
+* There is significant value in sharing a common engineering infrastructure for .NET Core3.
+* Buying into (not necessarily agreeing to always....grin) a shared/common infra is always hard - and expensive.
+* The 'how' Arcade does something is always open for discussion.  We should all be working together in an attempt to have the right implementation that works best for .NET Core 
+* We're all in the same boat - we all ship as part of .NET Core.  In other words, no 'privileged' repos here...  (the egalitarian sword cuts both ways though...)
```
Comment (jcagme) [inline Documentation/DefaultsGuidance.md]: Nit: These two are missing a '.' at the end

File: Documentation/DefaultsGuidance.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+* Weigh the costs of adopting different alternatives by looking at impact on existing Arcade repos across the product (.NET Core) as a whole.
+* Give consideration to how this affects new repos onboarding to Arcade.
+* We try and avoid blocking people based on differences, but still strongly encourage consistency.
+
+## Categories of Differences
+
+* Category 1 - Company, Divisional, or .NET Core requirements  
+* Category 2 - Differences which are truly exceptions to what is otherwise correct for .NET Core overall
+* Category 3 - Differences which are excessively expensive or risky to change.
+* Category 4 - Differences which are the result of team/individual preferences and/or history
```
Comment (jcagme) [inline Documentation/DefaultsGuidance.md]: Nit: mix of ',' ending and non-'.'

File: Documentation/DefaultsGuidance.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+* 'True' exceptions mean that there's a concrete business case driving the difference.  
+* These differences have 'defaults' (or preferences) that are set by Arcade which can be overridden **only** for this category of differences.
+* Preference or history don't fall into this category. 
+* These differences should always be the *exception*.  If multiple repos are experiencing the same differences, this likely points to a broader .NET Core (Arcade) change.
+
+### Category 3 - Differences which are excessively expensive or risky to change.
+
+* Sometimes, the 'right' thing to do is simply too expensive or risky to change right away.
+* 'Expensive' here is defined as taking more than 2 man-weeks of dev time to reconcile.
+* 'Risky' here is defined as likely destabilizing .NET Core  for more than 2 days.
```
Comment (jcagme) [inline Documentation/DefaultsGuidance.md]: >   [](start = 59, length = 2)

Nit: one space instead of two

**Comment by jcagme** [comment]
Mostly nits. Also, for folks using tools like CodeFlow would be great if lines are broken around column 100 so there is no need to scroll right and back :)

File: Documentation/ArcadeContributorGuidance.md (hunk 4 above)
Comment (markwilkie) [reply Documentation/ArcadeContributorGuidance.md]: Let's give 'em a pass for now

**Comment by markwilkie** [comment]
> * Turning an Arcade feature off or on -- perhaps reserve "opt out" and "opt in" for this e.g. opting out of an Arcade convention (test project discovery for example) and doing something explicitly (preferably for only a few projects and not entire repos of course)
> * Providing a value (input) for an Arcade feature -- some values may have defaults, some may not e.g. providing a repo-specific signing key

I was wanting (trying...grin) to stay away from implementation.  Do you think there's value in getting into more specifics for this doc?  The goal is to get agreement on the 'principles' and 'categories' so that subsequent discussions have some guidelines.  In other words, this doc (hopefully) helps us with the how to talk about how to deal with signing keys, but we still need to work through where the signing stuff lives.

Thoughts?

**Comment by markwilkie** [comment]
> Thanks for writing this up. I would add concrete examples of each category. Otherwise, LGTM!

I'm a little worried about getting into specific examples, as much context is usually "drug" into them.  Plus, I feel like most of these things can be placed into more than one category depending on how we think about it.  My hope is to establish the guidelines/principles for how we think about these things.  There's still work to do for each instance though....  Thoughts?

**Comment by dougbu** [comment]
> Do you think there's value in getting into more specifics for this doc?

@markwilkie I'm not sure I was looking for _additional_ details. Instead, I found the use of "opt in" and "opt out" to be confusing -- both are terms that usually center on whether a feature is on or off. Wanted different terms for "must provide per-repo values / inputs / whatever".

**Comment by markwilkie** [comment]
> I found the use of "opt in" and "opt out" to be confusing -- both are terms that usually center on whether a feature is on or off. Wanted different terms for "must provide per-repo values / inputs / whatever".

Ah - that makes sense.  Thanks @dougbu

File: Documentation/DefaultsGuidance.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+* The 'how' Arcade does something is always open for discussion.  We should all be working together in an attempt to have the right implementation that works best for .NET Core.
+* We're all in the same boat - we all ship as part of .NET Core.  In other words, no 'privileged' repos here...  (the egalitarian sword cuts both ways though...)
+* We prefer defaults that match the defaults used by 'vanilla' Visual Studio/C# code formatters/MSBuild/.NET Core SDK.
+* Weigh the costs of adopting different alternatives by looking at impact on existing Arcade repos across the product (.NET Core) as a whole.
+* Give consideration to how this affects new repos onboarding to Arcade.
+* We try and avoid blocking people based on differences, but still strongly encourage consistency.
+
+## Categories of Differences
+
+* Category 1 - Company, Divisional, or .NET Core requirements.
```
Comment (tmat) [inline Documentation/DefaultsGuidance.md]: Maybe it's already in Divisional category but I'd mention Visual Studio requirements as well.

File: Documentation/DefaultsGuidance.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+* Category 2 - Differences which are truly exceptions to what is otherwise correct for .NET Core overall.
+* Category 3 - Differences which are excessively expensive or risky to change.
+* Category 4 - Differences which are the result of team/individual preferences and/or history.
+
+## Category Criteria ("bar" for each) and Mitigation Guidance
+
+### Category 1 - Company, Divisional, or .NET Core requirements  
+
+* Hard requirements (like SDL or certs) that .NET Core must confirm to.
+* These differences **should have no defaults** and should have the required parameters explicitely provided by the repo.
```
Comment (tmat) [inline Documentation/DefaultsGuidance.md]: This point is not quite clear. I'd rather say:
If it is possible to comply with the requirements in multiple different ways, there should be no default in the SDK. However, the SDK should provide a set of options and repositories should choose from these given options rather then implementing a completely custom solution. It is still desirable that repositories are as uniform as possible with respect to these requirements.

File: Documentation/ArcadeContributorGuidance.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
+# Contributor Guidance for Arcade
+Over and above the 'normal' OSS contributor guidelines (which are important), there are some principles and guidelines that are specific to Arcade as well.
+
+For the most part, contributions to Arcade are straight forward and relatively 'smooth'.  However, from time to time, getting changes we want in can definitely be more challenging, and even frustrating.  This document attempts to clarify some of the expectations as well as provide some 'advice' for success when contributing to Arcade.
```
[the remark this answers is not in the document]
Comment (markwilkie) [reply Documentation/ArcadeContributorGuidance.md]: fixed

File: Documentation/DefaultsGuidance.md (hunk 9 above)
Comment (markwilkie) [reply Documentation/DefaultsGuidance.md]: done

File: Documentation/DefaultsGuidance.md (hunk 10 above)
Comment (markwilkie) [reply Documentation/DefaultsGuidance.md]: done

File: Documentation/ArcadeContributorGuidance.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
+# Contributor Guidance for Arcade
+Over and above the 'normal' OSS contributor guidelines (which are important), there are some principles and guidelines that are specific to Arcade as well.
```
[the remark this answers is not in the document]
Comment (markwilkie) [reply Documentation/ArcadeContributorGuidance.md]: added

File: Documentation/ArcadeContributorGuidance.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+For the most part, contributions to Arcade are straight forward and relatively 'smooth'.  However, from time to time, getting changes we want in can definitely be more challenging, and even frustrating.  This document attempts to clarify some of the expectations as well as provide some 'advice' for success when contributing to Arcade.
+
+## Principles
+* Should benefit multiple repos across .NET Core.
+* Every contributor is "heard" and taken seriously every time
+* Keep it simple. (don't over engineer or be "clever")
+* Focus on value over conviction.  (of course conviction should help drive the value discussion)
+* Opinions and beliefs must be substantiated with data and a strong business rationale.
+* All discussions (arguments) must be civil, respectful, productive, and above all - kind. The focus must always be on the topic, not individuals.
+* Work in 'partnership' (shoulder to shoulder) as apposed to 'us' vs. 'them'.
```
[the remark this answers is not in the document]
Comment (markwilkie) [reply Documentation/ArcadeContributorGuidance.md]: fixed

**Pull request merged**